bot/cogs/filter_lists.py

```diff
@@ -0,0 +1,250 @@
+import logging
+from typing import Optional
+
+from discord import Colour, Embed
+from discord.ext.commands import BadArgument, Cog, Context, IDConverter, group
+
+from bot import constants
+from bot.api import ResponseCodeError
+from bot.bot import Bot
+from bot.converters import ValidDiscordServerInvite, ValidFilterListType
+from bot.pagination import LinePaginator
+from bot.utils.checks import with_role_check
+
+log = logging.getLogger(__name__)
+
+
+class FilterLists(Cog):
+    """Commands for blacklisting and whitelisting things."""
+
+    methods_with_filterlist_types = [
+        "allow_add",
+        "allow_delete",
+        "allow_get",
+        "deny_add",
+        "deny_delete",
+        "deny_get",
+    ]
+
+    def __init__(self, bot: Bot) -> None:
+        self.bot = bot
+        self.bot.loop.create_task(self._amend_docstrings())
+
+    async def _amend_docstrings(self) -> None:
+        """Add the valid FilterList types to the docstrings, so they'll appear in !help invocations."""
+        await self.bot.wait_until_guild_available()
+
+        # Add valid filterlist types to the docstrings
+        valid_types = await ValidFilterListType.get_valid_types(self.bot)
+        valid_types = [f"`{type_.lower()}`" for type_ in valid_types]
+
+        for method_name in self.methods_with_filterlist_types:
+            command = getattr(self, method_name)
+            command.help = (
+                f"{command.help}\n\nValid **list_type** values are {', '.join(valid_types)}."
+            )
+
+    async def _add_data(
+            self,
+            ctx: Context,
+            allowed: bool,
+            list_type: ValidFilterListType,
+            content: str,
+            comment: Optional[str] = None,
+    ) -> None:
+        """Add an item to a filterlist."""
+        allow_type = "whitelist" if allowed else "blacklist"
+
+        # If this is a server invite, we gotta validate it.
+        if list_type == "GUILD_INVITE":
+            guild_data = await self._validate_guild_invite(ctx, content)
+            content = guild_data.get("id")
+
+            # Unless the user has specified another comment, let's
+            # use the server name as the comment so that the list
+            # of guild IDs will be more easily readable when we
+            # display it.
+            if not comment:
+                comment = guild_data.get("name")
+
+        # If it's a file format, let's make sure it has a leading dot.
+        elif list_type == "FILE_FORMAT" and not content.startswith("."):
+            content = f".{content}"
+
+        # Try to add the item to the database
+        log.trace(f"Trying to add the {content} item to the {list_type} {allow_type}")
+        payload = {
+            'allowed': allowed,
+            'type': list_type,
+            'content': content,
+            'comment': comment,
+        }
```

Comment:
I noticed that you mix single and double quotes. Is there a system to it?

Reply:
I'm afraid not - I mix them because single quotes are the prevailing style in all the repos at my job, and double quotes are the prevailing style here at pydis. Sometimes my fingers forget where I am.

I'll change these to double quotes.